### PR TITLE
print key and value on INI violations

### PR DESCRIPTION
**Pull request by cgzones**
On violations of INI settings include the key and if appropriate the value in the log message. This helps to locate offenders and fine tune the configuration itself.

### CODE REVIEW

File: src/sp_ini.c [hunk 1]
```diff
@@ -11,6 +11,33 @@
     sp_log_auto2("ini_protection", simulation, (cfg->policy_drop || (entry && entry->drop)), __VA_ARGS__); \
   }
 
+static const char* check_safe_key(const char* string) {
+  for (const char* p = string; *p != '\0'; p++) {
+    if (isalnum((unsigned char)*p) || *p == '_' || *p == '.') {
```
Comment (jvoisin) [inline src/sp_ini.c:16]: Don't forget to include `<ctype.h>` where `isalnum` is defined.

File: src/sp_ini.c [hunk 2]
```diff
@@ -11,6 +11,33 @@
...
+      continue;
+    }
+
+    return NULL;
```
Comment (jvoisin) [inline src/sp_ini.c:20]:
```suggestion
    if (!isalnum((int)*p) && *p != '_' && *p != '.') {
      return NULL;
    }
```

no need to do the whole loop if an invalid char is found. Also, `isalnum` takes an `int` as parameter, not an `unsigned char`.

File: src/sp_ini.c [hunk 3]
```diff
@@ -27,7 +54,7 @@ static bool /* success */ sp_ini_check(zend_string *const restrict varname, zend
   if (!entry) {
     if (cfg->policy_readonly) {
       if (!cfg->policy_silent_ro) {
-        sp_log_ini_check_violation("INI setting is read-only");
+        sp_log_ini_check_violation("INI setting `%s` is read-only", check_safe_key(ZSTR_VAL(varname)) ?: "<invalid>");
```
Comment (jvoisin) [inline src/sp_ini.c:57]: Looks like you're missing the left-hand side of the ternary here.

File: src/sp_ini.c [hunk 4]
```diff
@@ -66,14 +97,38 @@ static bool /* success */ sp_ini_check(zend_string *const restrict varname, zend
...
   if (entry->regexp) {
     if (!sp_is_regexp_matching_zstr(entry->regexp, new_value)) {
-      sp_log_ini_check_violation("%s", (entry->msg ? ZSTR_VAL(entry->msg) : "INI value does not match regex"));
+      if (entry->msg) {
+        sp_log_ini_check_violation("%s", ZSTR_VAL(entry->msg));
+      } else {
+        zend_string *base64_new_val = NULL;
+        const char *safe_new_val;
+
+        safe_new_val = check_safe_value(ZSTR_VAL(new_value));
```
Comment (jvoisin) [inline src/sp_ini.c:117]:
```suggestion
        const char *safe_new_val = check_safe_value(ZSTR_VAL(new_value));
```

**Review by jvoisin — CHANGES_REQUESTED** [review]

File: src/sp_ini.c (hunk 2 above)
Comment (cgzones) [reply src/sp_ini.c]: w.r.t. `int` vs `unsigned char` see https://wiki.sei.cmu.edu/confluence/display/c/STR37-C.+Arguments+to+character-handling+functions+must+be+representable+as+an+unsigned+char

Comment (jvoisin) [reply src/sp_ini.c]: It makes sense, thanks!

**Comment by jvoisin** [comment]
Nice idea and great implementation, thanks!

**Pull request merged**